guard/src/rules/eval_context.rs

```diff
@@ -1174,37 +1175,98 @@ impl<'value, 'loc: 'value> EvalContext<'value, 'loc> for RootScope<'value, 'loc>
+    UrlDecode,
+    ParseString,
+    ParseBoolean,
+    ParseFloat,
+    ParseInt,
+    ParseChar,
+}
+
+impl FunctionName {
+    pub fn get_expected_number_of_args(&self) -> usize {
+        match self {
+            FunctionName::Join => 2,
+            FunctionName::Substring | FunctionName::RegexReplace => 3,
+            FunctionName::Count
+            | FunctionName::JsonParse
+            | FunctionName::ToUpper
+            | FunctionName::ToLower
+            | FunctionName::UrlDecode
+            | FunctionName::ParseString
+            | FunctionName::ParseBoolean
+            | FunctionName::ParseFloat
+            | FunctionName::ParseInt
+            | FunctionName::ParseChar => 1,
         }
-    };
+    }
+}
 
-    if expected_num_args != num_args {
-        return Err(Error::ParseError(format!(
-            "{name} function requires {expected_num_args} arguments be passed, but received {num_args}"
-        )));
+impl std::fmt::Display for FunctionName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let name = match self {
+            FunctionName::Join => "join",
+            FunctionName::Substring => "substring",
+            FunctionName::RegexReplace => "regex_replace",
+            FunctionName::Count => "count",
+            FunctionName::JsonParse => "json_parse",
+            FunctionName::ToUpper => "to_upper",
+            FunctionName::ToLower => "to_lower",
+            FunctionName::UrlDecode => "url_decode",
+            FunctionName::ParseString => "parse_string",
+            FunctionName::ParseBoolean => "parse_boolean",
+            FunctionName::ParseFloat => "parse_float",
+            FunctionName::ParseInt => "parse_int",
+            FunctionName::ParseChar => "parse_char",
+        };
+        write!(f, "{}", name)
     }
+}
 
-    Ok(())
+impl TryFrom<&str> for FunctionName {
+    type Error = Error;
+
+    fn try_from(name: &str) -> std::result::Result<Self, Self::Error> {
+        match name {
+            "join" => Ok(FunctionName::Join),
+            "substring" => Ok(FunctionName::Substring),
+            "regex_replace" => Ok(FunctionName::RegexReplace),
+            "count" => Ok(FunctionName::Count),
+            "json_parse" => Ok(FunctionName::JsonParse),
+            "to_upper" => Ok(FunctionName::ToUpper),
+            "to_lower" => Ok(FunctionName::ToLower),
+            "url_decode" => Ok(FunctionName::UrlDecode),
+            "parse_string" => Ok(FunctionName::ParseString),
+            "parse_boolean" => Ok(FunctionName::ParseBoolean),
+            "parse_float" => Ok(FunctionName::ParseFloat),
+            "parse_int" => Ok(FunctionName::ParseInt),
+            "parse_char" => Ok(FunctionName::ParseChar),
+            _ => Err(Error::ParseError(format!(
+                "No function with the name '{name}' exists.",
+            ))),
+        }
+    }
 }
 
 // TODO: look into the possibility of abstracting functions into structs that all implement
```

Comment:
This `TODO` has got me wondering if we want to take this time to do just that. 

I.e. each FunctionName variant implements a trait called `Callable` with a function called `call` or something similar, and  then we handle the actual function calls there. Either that or we just move this function into an impl of the actual Enum.

What do you thin about that? Functionally it doesn't really change anything, and i supposed one might consider this to be abstracting for the sake of abstraction? Just my .02, if not we can just remove this TODO

Reply:
I think thats pretty cool - I can make that change